### PR TITLE
Added handling of selected entities for multistep display over AJAX

### PR DESCRIPTION
**Pull request by mtodor**
This is implementation of functionality described in ticket: https://www.drupal.org/node/2822009

### CODE REVIEW

File: js/entity_browser.multi_step_display.js [hunk 1]
```diff
@@ -20,9 +27,10 @@
       var $toggle = $('.entity-browser-show-selection');
 
       function setToggleText() {
-        if($entities.css('display') == 'none') {
+        if ($entities.css('display') == 'none') {
```
Comment (dawehner) [inline js/entity_browser.multi_step_display.js]: for better readability of the changeset its IMHO better to not include those unrelated changes, but rather add them to its own PR

File: js/entity_browser.multi_step_display.js [hunk 2]
```diff
@@ -53,4 +77,123 @@
...
+
+  /**
+   * Handle load entity over Ajax Request in selection display.
+   *
+   * @param {object} event
+   *   Event object.
+   * @param {Array} entity_ids
+   *   Entity Id that should be loaded into selection list.
+   */
+  Drupal.entityBrowserMultiStepDisplay.loadEntities = function (event, entity_ids) {
```
Comment (mtodor) [inline js/entity_browser.multi_step_display.js]: I think that, this should be decoupled more. Maybe best would be to go with one trigger for add/remove of entities and creating commands and then other trigger that will actually make execution of ajax request (and queued commands). That trigger can be used also to execute queue after running ajax request is finished. Also naming of methods should be improved to get things more understandable.

**Comment by mtodor** [comment]
There are few things that should be adjusted:
1. When View is without paging and there is auto select functionality enabled, then button "Select elements" is not displayed and selection is displayed after view, what makes it look ugly. There should be some padding there.
2. If "auto select" is enabled in configuration (for example yml file), but widget doesn't support it, then it should be checked in run-time that provided configuration is not functional.

EDIT:
1. This one is not relevant, since views are taking 100% width by default. If someone customize views, then it has to be done in their project. Simple:`.entities-list { clear: left; }` should be sufficient.

File: src/Form/EntityBrowserForm.php [hunk 3]
```diff
@@ -140,6 +143,29 @@ public function buildForm(array $form, FormStateInterface $form_state) {
...
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   Form status.
+   */
+  protected function isFunctionalForm(FormStateInterface $form_state) {
+    $current_widget_config = $this->entity_browser
+      ->getWidgets()
+      ->get($this->getCurrentWidget($form_state))
+      ->getConfiguration();
+
+    if ($current_widget_config['settings']['auto_select']) {
```
Comment (slashrsm) [inline src/Form/EntityBrowserForm.php]: This assumes internal configuration of the plugin. From shouldn't assume anything about the internals of the plugins. They should communicate just through functions defined on interfaces or events. Can we refactor this to work through a function?

**Comment by slashrsm** [comment]
Generally looks OK to me. Patch is big and I didn't manage to thoroughly review it, but the basic idea looks fine. Would be great if someone could do a proper review/testing.

File: modules/example/js/multistep-display-ajax.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+(function ($) {
+
+  'use strict';
+
+  /**
+   * Attaches the behavior of the media entity browser view.
+   */
+  Drupal.behaviors.multiStepAjaxSelection = {
+    attach: function () {
+      jQuery('.view-id-files_entity_browser_ajax td.views-field-filename')
```
Comment (mortenson) [inline modules/example/js/multistep-display-ajax.js]: This seems coupled to a specific view, and assumes that a table is used. Can we make this logic more generic?

File: modules/example/js/multistep-display-ajax.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        .once()
+        .click(function () {
+          var $row = $(this);
+          var $input = $row.find('input');
+
+          // Here is way, how to trigger loading of entity for multipstep
+          // display over ajax. On multistep display element, "load-entities"
+          // event should be triggered with value for entity that should be
+          // loaded. Value is in format: type:id - for example: file:3
+          $row.parents('form')
```
Comment (mortenson) [inline modules/example/js/multistep-display-ajax.js]: We shouldn't make other entity browsers implement this sort of logic - everyone using the View widget should benefit from the multistep work.

File: src/WidgetBase.php [hunk 6]
```diff
@@ -196,6 +205,14 @@ public function buildConfigurationForm(array $form, FormStateInterface $form_sta
       '#default_value' => $this->configuration['submit_text'],
     ];
 
+    // Allow "auto_select" setting when autoSelect is supported by widget.
```
Comment (mortenson) [inline src/WidgetBase.php]: Does auto-selection work when not using multi-step?

File: js/entity_browser.multi_step_display.js (hunk 2 above)
Comment (mortenson) [reply js/entity_browser.multi_step_display.js]: Agreed - I would prefer it each Widget broadcast JS events for adding and removing Entities, and this code simply waited for those events to happen and made AJAX calls when they did.

File: src/WidgetBase.php (hunk 6 above)
Comment (mtodor) [reply src/WidgetBase.php]: Currently - auto-selection only works in combination with multistep display, that's why I have added check in EntityBrowserForm.php to throw exception if not supported configuration is defined.

In general, it could be possible to make it in combination with other selection displays -> if you take a look in dropzonejs PR, you will see that widget form is actually submitted (same as it would be with button click), in that submit action it is possible to select entities, but after that request is finished, you have to rebuild EB form to get selection, or selection display part. And all this functionality is about -> not to rebuild form, just relevant parts.

File: src/Plugin/EntityBrowser/SelectionDisplay/MultiStepDisplay.php [hunk 7]
```diff
@@ -126,25 +137,171 @@ public function getForm(array &$original_form, FormStateInterface $form_state) {
...
+          'type' => 'fullscreen',
+        ],
+      ],
+    ];
+
     $form['use_selected'] = [
       '#type' => 'submit',
       '#value' => $this->t($this->configuration['select_text']),
       '#name' => 'use_selected',
-      '#access' => empty($selected_entities) ? FALSE : TRUE,
```
Comment (mtodor) [inline src/Plugin/EntityBrowser/SelectionDisplay/MultiStepDisplay.php:133]: This functionality should be fixed -> currently buttons are always displayed, but it should be displayed only when there is some selection.

So on AJAX request handling there should be following:
- on add: render and display buttons over AJAX commands (if selection was empty before)
- on remove: if selection is empty -> pass remove AJAX command to remove buttons

**Comment by slashrsm** [comment]
Will commit patch from the issue.
